File: AspNetMvcUnleashed/0672329980_source/Chapter04/VB/MvcApplication1/MvcApplication1/Helpers/ProductHelper.vb
Imports System.IO

Public Module ProductHelper

    <System.Runtime.CompilerServices.Extension()> _
    Function ProductList(ByVal helper As HtmlHelper) As String
        ' Get products from view data
        Dim products = CType(helper.ViewData.Model, IEnumerable(Of Product))

        ' Create HTML TextWriter
        Dim html = New HtmlTextWriter(New StringWriter())

        ' Open table
        html.RenderBeginTag(HtmlTextWriterTag.Table)

        ' Render product rows
        For Each product In products
            ' Open tr
            html.RenderBeginTag(HtmlTextWriterTag.Tr)

            ' Render name
            html.RenderBeginTag(HtmlTextWriterTag.Td)
            html.Write(product.Name)
            html.RenderEndTag()

            ' Render price
            html.RenderBeginTag(HtmlTextWriterTag.Td)
            html.Write("{0:c}", product.Price)
            html.RenderEndTag()

            ' Close tr
            html.RenderEndTag()
        Next 

        ' Close table
        html.RenderEndTag()

        Return html.InnerWriter.ToString()
    End Function


End Module

File: AspNetMvcUnleashed/0672329980_source/Chapter04/CS/MvcApplication1/MvcApplication1/Helpers/ProductHelper.cs
using System.Collections.Generic;
using System.IO;
using System.Web.Mvc;
using System.Web.UI;
using MvcApplication1.Models;

namespace MvcApplication1.Helpers
{
    public static class ProductHelper
    {
        public static string ProductList(this HtmlHelper helper)
        {
            // Get products from view data
            var products = (IEnumerable<Product>)helper.ViewData.Model;

            // Create HTML TextWriter
            var html = new HtmlTextWriter(new StringWriter());

            // Open table
            html.RenderBeginTag(HtmlTextWriterTag.Table);

            // Render product rows
            foreach (var product in products)
            {
                // Open tr
                html.RenderBeginTag(HtmlTextWriterTag.Tr);

                // Render name
                html.RenderBeginTag(HtmlTextWriterTag.Td);
                html.Write(product.Name);
                html.RenderEndTag();
 
                // Render price
                html.RenderBeginTag(HtmlTextWriterTag.Td);
                html.Write("{0:c}", product.Price);
                html.RenderEndTag();

                // Close tr
                html.RenderEndTag();
            }

            // Close table
            html.RenderEndTag();

            return html.InnerWriter.ToString();
        }


    }
}
